source/Octopus.Versioning.Tests/Octopus/OctopusVersionTests.cs

```diff
@@ -946,24 +982,27 @@ public void LargeVersionNumbersWillFail(string version)
 
         [Test]
         [TestCase("1.2.3-hi/there")]
+        [TestCase(" 1.2.3-hi/there ")]
         [TestCase("1.2.3-hi%there")]
+        [TestCase(" 1.2.3-hi%there ")]
         [TestCase("1.2.3-hi?there")]
+        [TestCase(" 1.2.3-hi?there ")]
         [TestCase("1.2.3-hi#there")]
+        [TestCase(" 1.2.3-hi#there ")]
         [TestCase("1.2.3-hi&there")]
+        [TestCase("1.2.3-hi there")]
         [TestCase(" ")]
         [TestCase("")]
         [TestCase(null)]
         public void IllegalCharsWillFail(string version)
         {
-            try
```

Comment:
I'm not entirely following why we modified this test?

Reply:
The changes here added more examples of versions that are considered invalid, and to also demonstrate how both the new parsing and old parsing agree that the versions are invalid. A number of the existing tests parse test cases with the new and old version parsers to show this relationship, and so this test was updated to do the same.

Reply:
But you said this used to work, from looking at history, this part of the code has never dealt with trimming inputs.
So the bug was introduced somewhere else, not here.
Hence IMO, this fix belongs to the consumers, not here.

Reply:
That is true, the clients should (and in our testing, do) trim versions. The root cause of the issue has been resolved elsewhere, this change is to make this class more "robust".

Reply:
I guess what I am saying is that IMO this class is correct and we should not be doing trims in it. It is up to the consumer to pass the correct input.
```c#
OctopusVersionParser.Parse() // fails if white spaces are passed
OctopusVersionParser.TryParse() // false if white spaces are passed
```
Same as any other class eg `Boolean.TryParse`.

I don't think this is the place to make the code more robust.

Reply:
Also, this would be a non-backwards change, which I have no idea what could break out there.

Reply:
Ok, I think I understand what you are trying to say.

I suspect this is where we will differ in opinion. Given the class that this one was designed to replace tolerated whitespace around versions, and such whitespace can be reliably ignored, I think it makes sense to change this class to both be more compatible with the previous class and to be more robust in the inputs it accepts.

This also speaks to your other question about what might break. Given we have already verified the clients (i.e. the web UI, cli, rest api, and client library) do strip spaces for these versions, the risk is low. Also this change actually increases comparability with any code that previously used classes like `SemVerFactory`.

Reply:
For some additional context, this issue has been discussed in slack, with members of the QRF team expressing a preference for making the version parsing more resilient: https://octopusdeploy.slack.com/archives/CNHBHV2BX/p1618527352494300?thread_ts=1618360491.423900&cid=CNHBHV2BX

Reply:
The regression is, that prior to this change: https://github.com/OctopusDeploy/OctopusDeploy/commit/e04a667c90de4fe278d1904b2092b2032fdab93e 
in: source/Octopus.Server/Web/Api/Components/DashboardBuilder.cs 

VersionFactory.CreateSemanticVersion("1.2.3 -Hello") would work, and return 1.2.3 as the version

But after this change: https://github.com/OctopusDeploy/Versioning/commit/e3d88e519c61ddb98a403b3428cbcf8d98185b18#diff-bed68acfe7ac983ac0f58e18c53eccc7f53061718be6d30f32dcb26551e7e471

It will now throw an exception... so _IF_ there is an invalid version in the database, it will now block functionality that previously worked.  This has been experienced by a customer already - and there may be more - especially if there was no data validation/migration performed for this change in requirements.

If a version can never have a space, then it should cost the same to strip or look for spaces. (btw Boolean.Parse(" true ") works)  I guess the problem arises if this parse function is also used for validation.  IMO parsing should be reasonably resilient, but validation should be strict... are we using this one function for both?  

I don't know how the client got an invalid version in the db, so until we do, or have verified the data as part of an update script, there remains the potential for it to happen again.  But it would seem that the previous parser was resilient to leading, trailing, and spaces before the hyphen:

![image](https://user-images.githubusercontent.com/81208241/114953373-98773200-9e9b-11eb-884b-334c4591f61c.png)

Reply:
Thanks, @mcasperson and @andrew-at-octopus for the extra context, this makes sense now.